scanner/src/main/kotlin/scanners/fossid/FossIdUrlProvider.kt

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright (C) 2022 The ORT Project Authors (see <https://github.com/oss-review-toolkit/ort/blob/main/NOTICE>)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.scanner.scanners.fossid
+
+import java.net.Authenticator
+import java.net.PasswordAuthentication
+
+import org.apache.logging.log4j.kotlin.Logging
+
+import org.ossreviewtoolkit.utils.common.replaceCredentialsInUri
+import org.ossreviewtoolkit.utils.common.toUri
+import org.ossreviewtoolkit.utils.ort.requestPasswordAuthentication
+
+/**
+ * An internal helper class that generates the URLs used by [FossId] to check out the repositories to be scanned.
+ *
+ * The URLs used by FossId can sometimes be different from the normal package URLs. For instance, credentials may need
+ * to be added, or a different protocol may be used. This class takes care of such mappings.
+ */
+internal class FossIdUrlProvider private constructor(
+    /**
+     * The URL mapping. URLs matched by a key [Regex] are replaced by the URL in the value. The replacement string can
+     * refer to matched groups of the [Regex]. It can also contain the variables [VAR_USERNAME] and [VAR_PASSWORD] to
+     * allow the injection of credentials.
+     */
+    private val urlMapping: Map<Regex, String>
+) {
+    companion object : Logging {
+        /** The prefix of option keys that define a URL mapping. */
+        internal const val PREFIX_URL_MAPPING = "urlMapping"
+
+        /** Variable in the mapping replacement string that references the username. */
+        private const val VAR_USERNAME = "#user"
```

Comment:
One last (optional) remark: You're documenting the variable as "username", you're calling the variable "USERNAME", but the marker is just "user". Should the full marker also be "#username" for consistency, and to reduce the likeliness that the shorter "#user" appears anywhere else in the URL?

Reply:
Yes, makes sense. Changed to the variable name "#username".